pkg/controller/velerobackup/hiveobject.go

```diff
@@ -0,0 +1,113 @@
+	logger   log.FieldLogger
+}
+
+func newHiveObject(object runtime.Object, logger log.FieldLogger) (*hiveObject, error) {
+	switch t := object.(type) {
+	case *hivev1.ClusterDeployment:
+		checksum := calculateChecksum(t.ObjectMeta, t.Spec, controllerutils.GetChecksumOfObjects, logger)
+		return &hiveObject{
+			checksum: checksum,
+			object:   object,
+			logger:   logger,
+		}, nil
+	case *hivev1.SyncSet:
+		checksum := calculateChecksum(t.ObjectMeta, t.Spec, controllerutils.GetChecksumOfObjects, logger)
+		return &hiveObject{
+			checksum: checksum,
+			object:   object,
+			logger:   logger,
+		}, nil
+	case nil:
+		return nil, fmt.Errorf("Can't convert nil to a HiveObject")
+	}
+
+	return nil, fmt.Errorf("Unknown Type: %T", object)
+}
+
+func (h *hiveObject) hasChanged() bool {
+	if h.checksum == errChecksum {
+		// The checksum function errored. In this case we assume the object has NOT changed.
+		return false
+	}
+
+	// Make sure newChecksum is different than the old checksum.
+	oldChecksum := h.getChecksumAnnotation()
+	return h.checksum != oldChecksum
+}
+
+func calculateChecksum(meta metav1.ObjectMeta, spec interface{}, checksumFunc controllerutils.ChecksumOfObjectsFunc, logger log.FieldLogger) string {
+	// We need to take the checksum WITHOUT the previous checksum
+	// to avoid an infinite loop of the checksum always changing.
+	metaCopy := meta.DeepCopy()
+	delete(metaCopy.Annotations, controllerutils.LastBackupAnnotation)
+
+	// We also need to take the checksum WITHOUT the previous ResourceVersion
+	// as when we update the checksum annotation, ResourceVersion changes.
+	metaCopy.ResourceVersion = ""
+
+	checksum, err := checksumFunc(metaCopy, spec)
+	if err != nil {
+		// We never return an error from this function as that could lead to a
+		// situation where other objects change, but because one of the object
+		// checksum errored, then the other objects won't get backed up.
+		// Instead we return a constant checksum value. This function should rarely,
+		// if ever, return an error.
+		logger.WithError(err).Info("error calculating object checksum")
+		return errChecksum
+	}
+
+	return checksum
+}
+
+func (h *hiveObject) getChecksumAnnotation() string {
+	tmpmeta, err := meta.Accessor(h.object)
+	if err != nil {
+		// This should never happen because we should always be an object.
+		panic(err)
+	}
+	annotations := tmpmeta.GetAnnotations()
+	return annotations[controllerutils.LastBackupAnnotation]
+}
+
+func (h *hiveObject) setChecksumAnnotation(checksum string) {
+	// Not checking err because we are guaranteed to be an object.
+	tmpmeta, _ := meta.Accessor(h.object)
+	annotations := tmpmeta.GetAnnotations()
+
+	if annotations == nil {
+		annotations = make(map[string]string)
+	}
+
+	// Actually set the checksum
+	annotations[controllerutils.LastBackupAnnotation] = checksum
+	tmpmeta.SetAnnotations(annotations)
+}
+
+func addHiveObjectLoggerFields(logger log.FieldLogger, hobj *hiveObject) *log.Entry {
```

Comment:
Why is this function not a member of hiveObject? If you do make it a member, then you can shorten the name to `addLoggerFields`.